viahtml/blocklist.py

```diff
@@ -46,15 +46,45 @@ def __init__(self, filename):
         self._filename = filename
         self._last_modified = None
         self.domains = {}
+        self.patterns = {}
 
         self._refresh()
 
     def is_blocked(self, url):
-        """Get the blocked reason for a URL, or None if it's not blocked."""
+        """Get the blocked reason for a URL, or False if it's not blocked."""
         self._refresh()
 
         domain = self._domain(url)
-        return self.domains.get(domain, False)
+        blocked = self.domains.get(domain)
+        if blocked:
+            return blocked
+
+        for pattern, reason in self.patterns.items():
+            if pattern.match(domain):
+                return reason
+
+        return False
+
+    def clear(self):
+        """Remove all domains from the blocklist."""
+        self.domains, self.patterns = {}, {}
+
+    def add_domain(self, domain, reason):
+        """Add a domain (or domain pattern) to the blocklist."""
+
+        reason = Blocklist.Reason.parse(reason)
+        if reason in self.PERMITTED:
+            # This is listed as blocked, but this service can actually
+            # serve this type without incident
+            return
+
+        if "*" in domain:
+            # Convert a string with '*' wildcards into a regex
+            pattern = "^" + re.escape(domain).replace("\\*", ".*") + "$"
+            pattern = re.compile(pattern, re.IGNORECASE)
+            self.patterns[pattern] = reason
+        else:
+            self.domains[domain] = reason
```

Comment:
This was added so we could separate loading the file from adding a domain. Previously we either had to load a file in the tests or mess about with internal variables.